Add reject_if option on expose_collection_of feature

Dhamsoft opened the pull request:
This option will determines if a record with the particular +attributes+ should be
  rejected by calling the reject_if Symbol or Proc (if defined).

  The reject_if option is defined by +expose_collection_of+.

nandahibatullah commented on lib/on_form/collection_wrapper.rb:
```diff
@@ -51,12 +54,12 @@ def parse_collection_attributes(params)
+            records_to_destroy << id.to_i if options[:allow_destroy]
           else
-            records_to_update[id.to_i] = attributes.except('id', :id, '_destroy', :destroy) if allow_update
+            records_to_update[id.to_i] = attributes.except('id', :id, '_destroy', :destroy) if options[:allow_update] && !call_reject_if(attributes)
```
could just have this in an if block as it's pretty long. Makes it a bit easier to read too.

nandahibatullah commented on lib/on_form/collection_wrapper.rb:
```diff
@@ -51,12 +54,12 @@ def parse_collection_attributes(params)
           end
         elsif !destroy
-          records_to_insert << attributes.except('_destroy', :destroy) if allow_insert
+          records_to_insert << attributes.except('_destroy', :destroy) if options[:allow_insert] && !call_reject_if(attributes)
```
same here

nandahibatullah commented on test/collection_form_test.rb:
```diff
@@ -106,6 +110,45 @@ def initialize(house)
+  describe "reject_if option as a symbol" do
+    it "ignores record if room_name is 'REJECT_CHECKER' on update" do
+      @room_listing_form.update!(
+        :rooms_attributes => {
```
here you have used two syntax of hash on line 6 new syntax and then here 116 are hash rockets, probs just stick to one

binary-koan commented on lib/on_form/collection_wrapper.rb:
```diff
@@ -2,17 +2,20 @@ module OnForm
 
-    def initialize(parent, association_name, collection_form_class, allow_insert, allow_update, allow_destroy)
+    def initialize(parent, association_name, collection_form_class, options = {})
+      default_options = {allow_insert: true, allow_update: true, allow_destroy: false, reject_if: nil}
```
I wonder if this would be cleaner with a DEFAULT_OPTIONS constant at the top of the class instead of a local variable?

binary-koan commented on lib/on_form/form.rb:
```diff
@@ -91,7 +95,13 @@ def self.expose_collection_of(association_name, on: nil, prefix: nil, suffix: ni
+          options.slice(:allow_insert, :allow_update, :allow_destroy, :reject_if)
+        )
+      end
+      # used by action_view's fields_for, and by the following lines
```
Can this be moved up above the method definition? Not clear to me what it refers to at the moment

binary-koan commented on test/collection_form_test.rb:
```diff
@@ -106,6 +110,45 @@ def initialize(house)
     proc { @rooms.first.reload}.must_raise ActiveRecord::RecordNotFound
   end
 
+  describe "reject_if option as a symbol" do
```
How about as a Proc / with different arity? IMO either we care about all those cases enough to add tests for all of them, or we don't care about them enough to have them at all (not sure how much we care about feature parity with ActiveRecord)

willbryant commented:
I'm not really keen on the way you've changed it back from modern kwargs to old style options hashes.  Why was that done?

Dhamsoft commented:
@willbryant @binary-koan 
`OnForm::Form.expose_collection_of` and `OnForm::CollectionWrapper.new` now accept keyword as arguments.

Dhamsoft committed:
Add reject_if option to expose_collection_of feature

  This option will determines if a record with the particular +attributes+ should be
  rejected by calling the reject_if Symbol or Proc (if defined).

  The reject_if option is defined by +expose_collection_of+.

 lib/on_form/collection_wrapper.rb | 34 +++++++++++++++++++++++++---------
 lib/on_form/form.rb               | 18 ++++++++++++++----
 test/collection_form_test.rb      | 73 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++-
 3 files changed, 111 insertions(+), 14 deletions(-)

Dhamsoft committed:
Update CollectionWrapper#new with modern keyword arguments format

lib/on_form/collection_wrapper.rb | 18 ++++++++----------
lib/on_form/form.rb               | 13 ++++++-------
2 files changed, 14 insertions(+), 17 deletions(-)

willbryant commented on lib/on_form/collection_wrapper.rb:
```diff
@@ -117,5 +117,19 @@ def add_errors_to_parent(parent_form, child_form)
+    # rejected by calling the reject_if Symbol or Proc (if defined).
+    # The reject_if option is defined by +expose_collection_of+.
+    def call_reject_if(attributes)
+      case callback = reject_if
```
?

willbryant commented on test/collection_form_test.rb:
```diff
@@ -3,7 +3,7 @@
   expose %i(street_number street_name city), on: :house
 
-  expose_collection_of :house_rooms, on: :house, as: :rooms, allow_insert: true, allow_update: true, allow_destroy: true do
+  expose_collection_of :house_rooms, :on => :house, :as => :rooms, :allow_insert => true, :allow_update => true, :allow_destroy => true do
```
Leftover from previous version?

Dhamsoft committed:
Remove unnecessary assignment and revert to new stle hash

Dhamsoft commented:
@willbryant  Removed unnecessary assignment and reverted back to new hash style 👍

willbryant commented:
Thanks!
